### ISSUE TITLE
Telescopic sums made effective

### ISSUE THREAD

**Issue by SnarkBoojum** [math-comp/math-comp#928]
From a discussion on zulip, telescopic sums bigop as-of-today mostly work to prove other telescopic sums results. They are too difficult to wield against specific examples:
```Coq
From mathcomp Require Import all_ssreflect all_algebra.
From mathcomp.algebra_tactics Require Import ring.

Set Implicit Arguments.
Unset Strict Implicit.
Unset Printing Implicit Defensive.

Import GRing.Theory Num.Theory.

Section First.

Local Open Scope ring_scope.

Variable RR: realFieldType.

(* that looks pretty bad already *)
Lemma test_without_telescope n: \sum_(1 <= k < n.+1) 1/(k * (k+1))%:R = 1-1/n.+1%:R :> RR.
Proof.
elim: n => [|n Hn].
  rewrite unlock //=.
  by field.
rewrite big_nat_recr //= Hn.
field. (* there field does its job well! *)
rewrite -[1 in X in _ && (X + _ != _)]/(1%:R) -!natrD.
by rewrite  -[0]/(0%:R) !eqr_nat.
Qed.

(* same bad-looking statement *)
Lemma test_with_telescope n: \sum_(1 <= k < n.+1) 1/(k * (k+1))%:R = 1-1/n.+1%:R :> RR.
Proof.
(* let's try to prepare the way *)
pose f: nat -> RR := fun k => (-1 / k%:R).
have simple_elt_decomp (k: nat): (1 <= k)%N -> f(k.+1) - f(k) = 1/ (k * (k + 1))%:R :> RR.
  move=> kneq0.
  unfold f.
  field.
  rewrite -[1 in X in (_ + X != _) && _]/(1%:R) -!natrD.
  rewrite -[0]/(0%:R) !eqr_nat.
  apply/andP ; split.
    by case k.
  by apply: lt0n_neq0.

under eq_big_nat => k /andP [kneq0 klen] do rewrite -simple_elt_decomp //=.
rewrite telescope_sumr.
unfold f.
field.
rewrite -[1 in X in (X + _ != _)]/(1%:R) -natrD -[0]/(0%:R) !eqr_nat //=.
by case n.
Qed.

End First.
```

I'm opening this bug report because @proux01 told me to propose a lemma to make things easier. I'll post a sample lemma when I'll have one, and when we'll be happy with it, open a PR.

**Comment by SnarkBoojum** [comment]
I'm not getting anywhere, stuck because I'm not yet fluent enough in Coq ; here is what I have:
```
From mathcomp Require Import all_ssreflect ssralg.

Set Implicit Arguments.
Unset Strict Implicit.
Unset Printing Implicit Defensive.

Section Missing.

Import GRing.Theory.
Local Open Scope ring_scope.

Lemma telescope_sumr_eq {a b: nat} {aleqb: a <= b} {V: zmodType} {u: nat -> V}:
 forall (f: nat -> V),
 (forall k, (a <= k < b)%N -> u k = f k.+1 - f k) -> \sum_(a <= k < b) u k = f b - f a.
Proof.
move=> f diffexpr.
under eq_big_nat => k kab do rewrite diffexpr //=.
by apply telescope_sumr.
Qed.

(* FIXME: here the idea was that I would want to say I want to use a telescopic sum, and coq
  would grant my wish only if I can prove the existence of an appropriate f - it doesn't quite
  work like this, even if I can prove this result *)
Lemma telescope_sumr_eq' {a b: nat} {aleqb: a <= b} {V: zmodType} {u: nat -> V}:
 (exists f, forall k, (a <= k < b)%N -> u k = f k.+1 - f k)
  -> exists f, \sum_(a <= k < b) u k = f b - f a.
Proof.
move=> [f diffexpr].
under eq_big_nat => k kab do rewrite diffexpr //=.
exists f.
by apply telescope_sumr.
Qed.

End Missing.

From mathcomp Require Import all_algebra.

Section Example.

Local Open Scope ring_scope.

Variable RR: realFieldType.

Lemma testing_telescope_sumr_eq n: \sum_(1 <= k < n.+1) 1/(k * (k+1))%:R = 1-1/n.+1%:R :> RR.
Proof.
Check telescope_sumr_eq. (* here I see the first argument is f *)
Check (telescope_sumr_eq _). (* but now it's the forall k, a <= k < b -> ... !? *)
apply (telescope_sumr_eq (fun k => -1/k%:R)). (* so this fails... *)

End Example.
```

**Comment by jouvelot** [comment]
You can move forward by using `rewrite` instead of `apply`:

rewrite ***@***.***_sumr_eq _ _ _ _  _ (fun k => -1/k%:R)).

**Comment by SnarkBoojum** [comment]
Here is a new version of the code ; as you can see I have:
- a working `telescope_sumr_eq` lemma ;
- an unsatisfying `telescome_sumr_eq'` lemma I would like to discuss ;
- an unsatisfying example of using `telescope_sumr_eq`: the first lines show the interest of the lemma, but there are so many awkward goals to take care afterwards... what can be done to improve that?

```
From mathcomp Require Import all_ssreflect ssralg.

Set Implicit Arguments.
Unset Strict Implicit.
Unset Printing Implicit Defensive.

Section Missing.

Import GRing.Theory.
Local Open Scope ring_scope.

Lemma telescope_sumr_eq {a b: nat} {V: zmodType} {u: nat -> V}:
 forall (f: nat -> V), (a <= b) ->
 (forall k, (a <= k < b)%N -> u k = f k.+1 - f k) -> \sum_(a <= k < b) u k = f b - f a.
Proof.
move=> f [aleqb diffexpr].
under eq_big_nat => k kab do rewrite diffexpr //=.
by apply telescope_sumr.
Qed.


(* FIXME: here the idea was that I would want to say I want to use a telescopic sum, and coq
  would grant my wish only if I can prove the existence of an appropriate f - it doesn't quite
  work like this, as I don't see how to express it with a single existence quantification.
  The lemma below is the best I managed to express&prove *)
Lemma telescope_sumr_eq' {a b: nat} {aleqb: a <= b} {V: zmodType} {u: nat -> V}:
 (exists f, forall k, (a <= k < b)%N -> u k = f k.+1 - f k)
  -> exists f, \sum_(a <= k < b) u k = f b - f a.
Proof.
move=> [f diffexpr].
under eq_big_nat => k kab do rewrite diffexpr //=.
exists f.
by apply telescope_sumr.
Qed.

End Missing.

(* FIXME: I have to do that so my arguments don't get optional under my feet
 -- I hope it won't bite me in the PR code *)
Arguments telescope_sumr_eq {a b V u} f.

From mathcomp Require Import all_algebra.
From mathcomp.algebra_tactics Require Import ring.

Section Example.

Import GRing.Theory Num.Theory.
Local Open Scope ring_scope.

Variable RR: realFieldType.

Lemma testing_telescope_sumr_eq n: \sum_(1 <= k < n.+1) 1/(k * (k+1))%:R = 1%:R - 1/n.+1%:R :> RR.
Proof.
rewrite (telescope_sumr_eq (fun (k: nat) => -1/k%:R)) //=.
  field.
  by rewrite -mulrS pnatr_eq0.
move=> k /andP [Hk0 Hkn].
field ; apply/andP ; split.
 - by rewrite pnatr_eq0 lt0n_neq0.
 - by rewrite -mulrS pnatr_eq0.
Qed.

End Example.
```

**Comment by jouvelot** [comment]
Note that, regarding `telescope_sumr_eq'`, it can in fact be decoupled from telescopic sums:
```
Lemma telescope_sumr_eq' {a b: nat} {aleqb: a <= b} {V: zmodType} {u: nat -> V}:
 (exists f, forall k, (a <= k < b)%N -> u k = f k.+1 - f k)
  -> exists f, \sum_(a <= k < b) u k = f b - f a.
Proof.
move=> _.
exists (fun x => if a == b then 0 else if x == b then \sum_(a <= k < b) u k else 0).
have [] := boolP (a == b) => [/eqP ->|]; last by rewrite eq_refl subr0.
rewrite subr0 big_nat big_pred0 // => x.
by rewrite leqNgt andNb.
Qed.
```

**Comment by jouvelot** [comment]
Indeed, the statement `telescope_sumr_eq'` can be proven without referring to telescopic sums whatsoever. It seems to me that `telescope_sumr_eq` is already the good one. As for the remaining trivial goals to prove in the example, at least, they are.... "trivial" :)

**Comment by SnarkBoojum** [comment]
Their proof is definitely not simple enough compared to their statement -- I tend to consider that a bug.

**Comment by jouvelot** [comment]
"Bug" seems to me a bit too strong a word: "inconvenience" could be better. But your mileage may vary.... ;)

**Comment by SnarkBoojum** [comment]
Well, the target I have in mind is a generic first-year student, so "inconvenience" is an understatement... each of those `rewrite` line and the `apply/andP` are pretty terrible!

**Comment by proux01** [comment]
@SnarkBoojum 

For `telescope_sumr_eq`, I would probably put the arguments in the following order (f is the thing you usually want to provide, let's put it as early as possible):
```Coq
From mathcomp Require Import all_ssreflect all_algebra.
From mathcomp.algebra_tactics Require Import ring.

Set Implicit Arguments.
Unset Strict Implicit.
Unset Printing Implicit Defensive.

Import GRing.Theory Num.Theory.

Local Open Scope ring_scope.

Section First.

Lemma telescope_sumr_eq (V : zmodType) (f u : nat -> V) (a b: nat) :
    (a <= b)%N -> (forall k, (a <= k < b)%N -> u k = f k.+1 - f k) ->
  \sum_(a <= k < b) u k = f b - f a.
Proof.
move=> aleqb diffexpr.
under eq_big_nat => ? ? do rewrite diffexpr //.
exact: telescope_sumr.
Qed.

End First.
Arguments telescope_sumr_eq {V} f u.
```
For the use, it looks good to me on your example:
```Coq
Section Test1.

Variable F : realFieldType.

Lemma testing_telescope_sumr_eq n :
  \sum_(1 <= k < n.+1) 1 / (k * k.+1)%:R = 1%:R - 1 / n.+1%:R :> F.
Proof.
rewrite (telescope_sumr_eq (fun k => - 1 / k%:R)) => [|//|k /andP[kgt0 _]].
  by field; rewrite -mulrS pnatr_eq0.
by field; rewrite -mulrS !pnatr_eq0 -lt0n kgt0.
Qed.
```
(although, as already discussed, it would be nice to have some automation to solve the goals with `%:R` left by field)

> Well, the target I have in mind is a generic first-year student, so "inconvenience" is an understatement... each of those rewrite line and the apply/andP are pretty terrible!

Wow, that seems ambitious to me, maybe you can discuss with @mkerjean and @amahboubi who, IIRC, teach an introductory Coq lecture for early math students.

**Comment by ybertot** [comment]
@proux01 For your information, I am also involved in this project and discussing with @SnarkBoojum.  [This site](https://liberabaci.gitlabpages.inria.fr/) may help understanding some of our reflection on the matter.

**Comment by amahboubi** [comment]
Hello @SnarkBoojum . Usually, using more forward chaining steps (with `have`, `suffices`, `pose`, `set`, ...) helps writing more satisfactory proof scripts.

**Comment by SnarkBoojum** [comment]
I finally dived again in this ; I also wrote a `telescope_prodr_eq` because that's definitely useful on a few occasions too:
```
From mathcomp Require Import all_ssreflect ssralg.

Set Implicit Arguments.
Unset Strict Implicit.
Unset Printing Implicit Defensive.

Section Missing.

Import GRing.Theory.
Local Open Scope ring_scope.

Lemma telescope_sumr_eq {a b: nat} {V: zmodType} {u: nat -> V} {aleqb: a <= b}:
 forall f: nat -> V,
 (forall k, (a <= k < b)%N -> u k = f k.+1 - f k)
 -> \sum_(a <= k < b) u k = f b - f a.
Proof.
move=> f diffexpr.
under eq_big_nat => k kab do rewrite diffexpr //=.
by apply: telescope_sumr.
Qed.

Lemma telescope_prodr_eq {a b: nat} {R: comUnitRingType} {u: nat -> R} {altb: a < b}:
 forall (f: nat -> R),
   (forall k, a< k < b -> f k \is a GRing.unit)
   -> (forall k, (a <= k < b)%N -> u k = f k.+1 / f k)
   -> \prod_(a <= k < b) u k = f b / f a.
Proof.
move=> f Hf quotexpr.
under eq_big_nat => k kab do rewrite quotexpr //=.
rewrite (telescope_big (fun i j => f j / f i)) ?altb //.
move=> k kab //=.
rewrite mulrC mulrA divrK //.
by exact: Hf k kab.
Qed.

End Missing.

(* FIXME: I have to do that so my arguments don't get optional under my feet
 -- I hope it won't bite me in the PR code *)
Arguments telescope_sumr_eq {a b V u aleqb} f.
Arguments telescope_prodr_eq {a b R u altb} f.

From mathcomp Require Import all_algebra.
From mathcomp.algebra_tactics Require Import ring.

Section Example.

Import GRing.Theory Num.Theory.
Local Open Scope ring_scope.

Variable RR: realFieldType.

Lemma testing_telescope_sumr_eq n:
  \sum_(1 <= k < n.+1) 1/(k * (k+1))%:R = 1%:R - 1/n.+1%:R :> RR.
Proof.
rewrite (telescope_sumr_eq (fun (k: nat) => -1/k%:R)) //=.
  field.
  by rewrite -mulrS pnatr_eq0.
move=> k /andP [Hk0 Hkn].
field.
apply/andP ; split.
 - by rewrite pnatr_eq0 lt0n_neq0.
 - by rewrite -mulrS pnatr_eq0.
Qed.

Lemma testing_telescope_prodr_eq n:
  n != 0%N
  -> \prod_(1 <= k < n.+1) (k.+1%:R / k%:R) = n.+1%:R :> RR.
Proof.
move=> nneq0.
rewrite (telescope_prodr_eq (fun (k: nat) => k%:R)) //=.
- by field.
- move: nneq0.
  by case: n.
- move=> k /andP [kgt1 _].
  apply: unitf_gt0.
  rewrite ltr0n.
  by apply: ltn_trans _ kgt1.
Qed.

End Example.
```

How does it look?

**Comment by proux01** [comment]
Seems good for a PR. As I said above, I would change the order of the arguments:
```Coq
Lemma telescope_sumr_eq {a b : nat} {V : zmodType} (f u : nat -> V) :
   (a <= b)%N ->
   (forall k, (a <= k < b)%N -> u k = f k.+1 - f k) ->
 \sum_(a <= k < b) u k = f b - f a.
```
But let's discuss those details in the PR.

**Comment by SnarkBoojum** [comment]
@proux01 I wanted most arguments to be optional (and put them first) and only then `f` and the fact it was a suitable candidate for telescoping as final mandatory ones, but that's just because I tend to distinguish between "stupid" hyps and active ones in my mathematical statements and hence mimick that in Coq.

That brings two questions:
1. what would the other order bring in?
2. why is it better to discuss in a PR and not here?

**Comment by proux01** [comment]
>     1. what would the other order bring in?

It's easier to provide `u` in case you want to rewrite backward.

>     2. why is it better to discuss in a PR and not here?

Because we have an actual code (rather than a bunch of messages) and github offers good tools to visualize diffs and make suggestions.

**Comment by SnarkBoojum** [comment]
Ok, PR done.

**Comment by proux01** [comment]
Addressed in #942
